### PR TITLE
Progress bar for upload/download files

### PR DESCRIPTION
**Pull request by svartkanin**
I added a progress bar for the uploaded/downloaded files and also adjusted it to the used terminal size. Furthermore I also added a little additional output stating the total of files that are going to be synchronized

### CODE REVIEW

**Comment by vitalif** [comment]
- code style
- CurlAgent is not the right place for printing progress bars...
- m_downloadFileBytes should be just m_size

**Comment by svartkanin** [comment]
Which code style did you use so I can adjust it? I used K&R should I change it to GNU?
In which other part would you like to move the printing of the progress bar?
And isn't m_size a bit to broad, since the content is only the download size not upload!?

**Comment by vitalif** [comment]
I like both curly braces to be on separate lines, but that's not the main point - at least it should be consistent, and parts of patch here https://github.com/vitalif/grive2/pull/90/commits/d9e925cf3ce0c97de05943aaa8531df97c68623d have different style (tabs/spaces/different size of indentation/different brace style); also it should be consistent with the surrounding code. Also there is a typo: totalDownlaodSize :)

progress bar printing and terminal size determination should be done in some separate place, probably in some additional class in util/ so it can be plugged in or out. CurlAgent should just call a progress callback (with total/done bytes) passed from outside (not global).

Entry represents a server-side file, so m_size is ok because it's the server-side file size.

Also, in fact, I have another question: isn't it possible to print progress fully without passing size parameter into http agent from outside? I.e. http client 100% knows upload size when uploading and it probably sees Content-Length header when downloading. So it already has all the needed information to report progress (done/total bytes).

And this magic also shouldn't be there in CurlAgent:

```
+               if (url.compare("https://accounts.google.com/o/oauth2/token"))
+                       progressBar = true;
+               else
+                       progressBar = false;
+       } else {
+               if (!boost::starts_with(url, "https://www.googleapis.com/")) {
+                       progressBar = true;
+                       totalDownlaodSize = downloadFileBytes;
+               } else
+                       progressBar = false;
```

**Comment by cinquecento** [comment]
Beside the units of data transferred don't show
```
  [ 74%] [====================================>             ] 907.27/1.19GB
  [ 84%] [=========================================>        ] 1019.80/1.19GB
  [ 97%] [===============================================>  ] 1.16/1.19GB
  [100%] [=================================================>] 1.19/1.19GB
```
for comparison, if the units show
```
  [ 74%] [====================================>             ] 907.27MB/1.19GB
  [ 84%] [=========================================>        ] 1019.80MB/1.19GB
  [ 97%] [===============================================>  ] 1.16GB/1.19GB
  [100%] [=================================================>] 1.19GB/1.19GB
```
should the "Progress bar" option be exclusively enabled by e.g. --progress-bar ?

**Comment by svartkanin** [comment]
I'll add a new progress-bar parameter, if vitalif agrees with it. 
I haven't come around to work on it that much lately but I'll take care of it

**Comment by vitalif** [comment]
+1, an option would be good

**Comment by svartkanin** [comment]
Alright, I moved the progress bar to a new class. There is also a new parameter which allows the enabling of the progress bar. 

I have tried to work without passing the size parameter. The callback function which is given to curl is only called when uploading files properly. But when downloading files the function either called with a zero value in the total download size parameter or not at all. I tested it with different files, and when having files that contain non readable data (e.g. byte streams) curl doesn't seem to be able to determine the download size. 
I also looked into the header callback function, which should contain the content-length but it didn't. 
Therefore I didn't see any other way in getting the download size within the client without having to pass the additional parameter. 

Let me know if you require any further changes

**Comment by cinquecento** [comment]
```
$ grive --version
grive version 0.5.1-dev Sep 13 2016 15:29:37

$ grive --help | grep bar
  --progress-bar              Enable progress bar for upload/download of files

$ grive --progress-bar --dir Test/
Reading local directories
Reading remote server file list
1 total changes
Synchronizing files
sync "./Test/10MB.test" doesn't exist in server, uploading
  [  5%] [=>                                              ] 496.00KB/10.00MB
  [ 11%] [====>                                           ] 1.06MB/10.00MB
  [ 33%] [===============>                                ] 3.30MB/10.00MB
  [ 77%] [====================================>           ] 7.72MB/10.00MB
  [ 99%] [==============================================> ] 9.89MB/10.00MB
  [100%] [===============================================>] 10.00MB/10.00MB
Finished!
```
Magnífico!

**Comment by cinquecento** [comment]
There is another, let's say the issue;

With Maximizing / Unmaximizing i.e. resizing a terminal window - [xfce4-terminal](https://git.xfce.org/apps/xfce4-terminal/) ...
```
$ stty size
24 80
$ stty size
38 146
$ stty size
24 80
$ stty size
38 146
```
... output becomes a bit messy
```
$ grive --progress-bar
Reading local directories
Reading remote server file list
3 total changes
Synchronizing files
sync "./.../kernel-4.7.4-234.fc24.src.rpm" doesn't exist in server, uploading
  [ 51%] [==================================================>                     [ 93%] [============================================>   ] 80.48MB/86.71MB
===============================================>                                                ] 44.91MB/86.71MB
```

**Comment by svartkanin** [comment]
I have noticed that "issue", but I think that behavior is actually part of the terminal.
I have looked into other possibilities in printing the progress bar, like boost which gives a limited amount of flexibility and also ncursor which actually works quite fine, but clear the entire screen every time when performing an output...
So this is the best way I could think of performing it with the information outputted. 

But I'm open for suggestions :)

**Comment by f2404** [comment]
Hi, I'm xfce4-terminal developer, @cinquecento has reached me on this topic.

First, I guess the issue is not specific to xfce4-terminal but rather to vte widget the terminal is based on. vte is also used by other terminal apps such as gnome-terminal, lxterminal, etc. You can run `vte-2.91` app and check the issue in it.
Second, when running `wget file` in `vte-2.91` and resizing the window I'm seeing a similar issue - when shrinking the window. This makes me think that aligning progress bar length to changing terminal width is not a straight-forward thing in general.

However, I would suggest trying your scenario in xterm (which is kind of a reference for terminal apps). If its behavior is better than vte's, it will be worth opening an issue against vte: https://bugzilla.gnome.org/

**Comment by svartkanin** [comment]
I've tried xterm now and there seems to be a better behavior, can you confirm that @cinquecento?

File: libgrive/src/protocol/AuthAgent.cc [hunk 1]
```diff
@@ -48,6 +48,11 @@ void AuthAgent::SetLog( http::ResponseLog *log )
 	return m_agent->SetLog( log );
 }
 
+void AuthAgent::SetProgressBar( ProgressBar *progressbar )
+{
+	return m_agent->SetProgressBar( progressbar );
```
Comment (f2404) [inline libgrive/src/protocol/AuthAgent.cc]: The method is `void`; why does it return something?

File: grive/src/main.cc [hunk 2]
```diff
@@ -149,13 +151,16 @@ int Main( int argc, char **argv )
...
 	Config config(vm) ;
-	
+
 	Log( "config file name %1%", config.Filename(), log::verbose );
 
 	std::unique_ptr<http::Agent> http( new http::CurlAgent );
 	if ( vm.count( "log-http" ) )
 		http->SetLog( new http::ResponseLog( vm["log-http"].as<std::string>(), ".txt" ) );
 
+	ProgressBar *pb = new ProgressBar(vm.count( "progress-bar" ) != 0);
```
Comment (f2404) [inline grive/src/main.cc]: Where does this object get deleted?

File: libgrive/src/http/Agent.hh [hunk 3]
```diff
@@ -59,7 +60,8 @@ public :
 	virtual long Get(
 		const std::string& 	url,
 		DataStream			*dest,
-		const Header&		hdr ) ;
+		const Header&		hdr,
+		const long			downloadFileBytes = 0) ;
```
Comment (f2404) [inline libgrive/src/http/Agent.hh]: `const` has no sense here as function cannot modify this parameter anyway.

**Comment by cinquecento** [comment]
@svartkanin that's right, xterm refreshes the screen contents.

**Comment by vitalif** [comment]
Sorry, I didn't have much time to review it yet.
But I don't think bad ascii progress bar display after terminal resizing is a big issue. 
Although it should be not really hard to fix by checking terminal size occasionally.
```struct winsize size;
ioctl(STDOUT_FILENO,TIOCGWINSZ,&size);```

**Comment by vitalif** [comment]
Sorry, I've meant
> bad ascii progress bar display after terminal resizing is NOT a big issue

**Comment by cinquecento** [comment]
[Rebase to 20160928gitdd77c99](https://paste.fedoraproject.org/436953/raw) - [svartkanin/grive2](https://github.com/svartkanin/grive2)

**Comment by vitalif** [comment]
@svartkanin do you want change git identity or this should be committed as `username <you@example.com>` or `Blackrabbit <example@example.com>`? :)

**Comment by vitalif** [comment]
for reference, this is the rest of unneeded/code style changes that I've reverted: https://github.com/vitalif/grive2/commit/33c39a5c6ea753aa2bc7202cd5732816a8afde2b

**Comment by vitalif** [comment]
I'm working on this patch in the `progressbar` branch.
Another note: Entry::Size() should of course be `u64_t`, not 32-bit `long`. However it doesn't matter anymore because Entry already has Size() method after implementing #101.

**Comment by vitalif** [comment]
One more note: what's the idea of `getTotalChanges()` method that returns total number of **files**, not **changes**?

**Comment by svartkanin** [comment]
Sorry I was not able to check on it the last couple of days...
So lets see, are the issues mentioned by @f2404 already addressed, otherwise I can include those as well?
@vitalif the method `getTotalChanges()` should list the total files changed at the beginning of the synchronization

P.S.: I'll try to edit the identity

**Comment by vitalif** [comment]
> @vitalif the method getTotalChanges() should list the total files changed at the beginning of the synchronization

Now it doesn't list total changed files, it just prints the number of ALL files. I've removed that feature - make it work and create a separate pull request if you want it.

I've reworked your patch, newer version is in `progressbar` branch (https://github.com/vitalif/grive2/commits/progressbar). It's ready to be merged, say what identity you want and I'll merge it :)

**Comment by svartkanin** [comment]
Sorry I've missed that...
You can just put Blackrabbit (I'm not on a computer right now to change it :) )

**Comment by vitalif** [comment]
> You can just put Blackrabbit (I'm not on a computer right now to change it :) )

without email?

**Comment by svartkanin** [comment]
Sorry, I updated the account info now, hope you can see it, otherwise it's blackrabbit256@gmail.com

**Comment by vitalif** [comment]
Ok, merged.

**Comment by cinquecento** [comment]
See https://github.com/vitalif/grive2/issues/103

**Comment by vitalif** [comment]
@f2404 @cinquecento About terminal resizing: it works without problem when you run grive directly in a terminal, because util/ProgressBar.cc calls TIOCGWINSZ ioctl every time before printing progressbar. There were some glitches in xfce4-terminal caused by the `clear whole line` escape sequence (\33[2K) printed before progressbar - I've replaced it wirh \33[K (clear to the end of line) after progressbar, and it's smooth now.

However, when you run grive in a subshell of Midnight Commander, terminal resizing is broken. But that's probably a bug of mc subshell, not the terminal emulator itself of course.

**Pull request closed without merging**